### PR TITLE
CLOUDSDK-165

### PR DESCRIPTION
**Pull request by philipp-heyse**

### CODE REVIEW

File: lib/actions/BackendAction.js [hunk 1]
```diff
@@ -90,6 +93,16 @@ class BackendAction {
     })
   }
 
+  async writeExtensionConfigs () {
+    const attachedExtensions = this.appSettings.loadAttachedExtensions()
+    const promises = []
+    for (let extensionId in attachedExtensions) {
+      const config = attachedExtensions[extensionId]
```
Comment (pvomhoff) [inline lib/actions/BackendAction.js]: The content of the `config` is not the value of the `extension-config`
You need to read it separately.

File: lib/actions/BackendAction.js [hunk 2]
```diff
@@ -45,13 +45,19 @@ class BackendAction {
     const pid = await utils.previousProcess('backend', this.settingsFolder)
     if (pid) throw new Error(`Backend process is already running with pid: ${pid}. Please quit this process first.`)
 
-    this.dcClient = new DcHttpClient(this.userSettings, logger)
+    this.dcClient = this.dcClient
+      ? this.dcClient
```
Comment (pvomhoff) [inline lib/actions/BackendAction.js]: Why do you check if the dcClient is already exisit?
The run function is the first function that get called when an action from the terminal is dispatched

File: lib/actions/BackendAction.js [hunk 3]
```diff
@@ -90,6 +96,18 @@ class BackendAction {
     })
   }
 
+  async writeExtensionConfigs () {
+    const attachedExtensions = this.appSettings.loadAttachedExtensions()
+    const promises = []
+    for (let extensionId in attachedExtensions) {
+      const extensionsFolder = path.join(this.appSettings.getApplicationFolder(), AppSettings.EXTENSIONS_FOLDER)
```
Comment (pvomhoff) [inline lib/actions/BackendAction.js]: You can move this line out of the for loop

File: test/actions/BackendAction-test.js [hunk 4]
```diff
@@ -393,15 +393,58 @@ describe('BackendAction', () => {
...
+        mockConf,
+        (err) => {
+          assert.ifError(err)
+          let called = 0
+          backendAction.appSettings.getId = () => {
+            return 1
+          }
+          backendAction.dcClient.generateExtensionConfig = (config) => {
+            called++
+            console.log(config)
```
Comment (pvomhoff) [inline test/actions/BackendAction-test.js]: console.log!

File: test/actions/BackendAction-test.js [hunk 5]
```diff
@@ -393,15 +393,58 @@ describe('BackendAction', () => {
...
+          }
+
+          appSettings.loadAttachedExtensions = () => { return {testExtension: {path: 'test-extension'}} }
+          backendAction._startSubProcess = () => {}
+
+          backendAction.run({}).then(() => {
+            assert.ok(called !== 0, 'dcClient generateExtensionConfig should have been called at least once')
+            done()
+          }).catch(err => {
+            console.log(err)
```
Comment (pvomhoff) [inline test/actions/BackendAction-test.js]: console.log

File: lib/actions/BackendAction.js (hunk 2 above)
Comment (philipp-heyse) [reply lib/actions/BackendAction.js]: moved init to constructor, needs to be changeable for unit test

File: test/actions/BackendAction-test.js [hunk 6]
```diff
@@ -393,15 +393,56 @@ describe('BackendAction', () => {
...
+        'test-extension',
+        'extension-config.json')
+
+      fsEx.createFileSync(extensionConfigPath)
+
+      let mockConf = {someKey: 'someVal'}
+      fsEx.writeJson(
+        extensionConfigPath,
+        mockConf,
+        (err) => {
```
Comment (Rekhyt) [inline test/actions/BackendAction-test.js]: As discussed this test could probably be done using async/await to be a bit more readable.

File: test/actions/BackendAction-test.js [hunk 7]
```diff
@@ -393,15 +393,56 @@ describe('BackendAction', () => {
...
         assert.ifError(err)
       }
     })
 
+    it('should write the extension configs for the app', done => {
+      backendAction.appSettings.EXTENSIONS_FOLDER = 'extensions'
+      const extensionConfigPath = path.join(
+        backendAction.appSettings.getApplicationFolder(), AppSettings.EXTENSIONS_FOLDER,
+        'test-extension',
+        'extension-config.json')
```
Comment (Rekhyt) [inline test/actions/BackendAction-test.js]: Closing parenthesis should go into the next line.

**Pull request merged**